### ISSUE TITLE
HTML, JS and CSS content inside Nostr

### ISSUE THREAD

**Issue by Marcelo-Theodoro** [nostr-protocol/nips#298]
I've been exploring how could we have entire web applications running inside Nostr: https://github.com/Marcelo-Theodoro/nnterra

The work is still in the early stages, but I'd love to know what you think about a NIP defining a [Parameterized Replaceable Event](https://github.com/nostr-protocol/nips/blob/master/33.md) for contents with complex code, like the one in the title.

The motivations are:
- Currently, we don't have any other kind for this type of content
- Relays may choose to NOT store this kind of event since it can take more space
- Developers can build complex web apps and frameworks to have both the web applications and their content "censorship resistant"

**Comment by staab** [comment]
That's a no from me dawg. The security holes this opens up are wider than I am tall. A sandboxed alternative like Lua could work, but would require a lot of work for any supporting client.

**Comment by Marcelo-Theodoro** [comment]
> security holes this opens

Could you specify? There's any security hole this open up that we don't already have with common web applications?

> sandboxed alternative like Lua could work

That's was my initial plan. I didn't go forward with that because was not clear to me the advantages of sandboxed vs. the "default" stack for web.

**Comment by staab** [comment]
There was a nostr app back in December that had a vulnerability to this type of attack (called [HTML injection](https://www.invicti.com/learn/html-injection/) and was shut down as a result.

**Comment by Marcelo-Theodoro** [comment]
> There was a nostr app back in December that had a vulnerability to this type of attack (called [HTML injection](https://www.invicti.com/learn/html-injection/) and was shut down as a result.

anigma, wasn't?

I agree that this is a concern! I believe the point in question is: a webapp stored on a relay using Nostr protocol would have a bigger risk than any other web application?

**Comment by staab** [comment]
Yes, because it's embedded in a web application that has access to your private key (or at least a signing function)

**Comment by Marcelo-Theodoro** [comment]
I don't really think we are talking about the same thing.

I'm not talking about embedding user posts in some other web application.
I'm talking a layer above: Entire web applications stored on relays.

See: The web application `example.com/app` is deployed on Firebase.
When you type `example.com/app` in the conventional browser, it'll:
1 - open an HTTP connection with the firebase server
2 - download the resources related to that URL
3 - finally, render the content
*(there are a lot more steps, actually. But it doesn't matter for the example)*

What I have in mind is the same, but instead of storing code in Firebase, you could store in multiple Nostr relays.

Example:
You type `nostr://{pubkey}/{event-id}/?relay=example1.com&relay=example2.com` in some kind of specialized "Nostr browser".
The "Nostr browser" will:
1 - open a connection with `example1.com` and `example2.com` 
2 - send a REQ for `{event-id}`
3 - once one of the relays answer, it will validate if the pubkey is the same as `{pubkey}` and has a valid `sig`
4 - finally, it will render the content from `event` (maybe even doing extras REQ for other resources if required)

Note that `nostr://{pubkey}/{event-id}/?relay=example1.com&relay=example2.com` and `example.com/app` could be the exactly same application.

Of course, there are risks. My point is that it doesn't matter if the application is stored on Firebase, or Nostr. The risks are the same.

**Comment by hendore** [comment]
@Marcelo-Theodoro Have you come across the interplanetary filesystem? https://ipfs.tech/ Brave currently supports IPFS out the box, other browsers can handle the IPFS protocol with an extension.

**Comment by staab** [comment]
I see, yes I was misunderstanding. Interesting idea, but I'm not sure how useful it would be.

**Comment by Marcelo-Theodoro** [comment]
> @Marcelo-Theodoro Have you come across the interplanetary filesystem? https://ipfs.tech/ Brave currently supports IPFS out the box, other browsers can handle the IPFS protocol with an extension.

I did not. I'll take a look.

> I see, yes I was misunderstanding. Interesting idea, but I'm not sure how useful it would be.

Cool! Neither do I haha
Didn't find a great use case yet.

**Comment by rusteroobs** [comment]
There is an editor written in jsx, I wonder if nostr could not be integrated into the editor. [blocks](https://github.com/blocks) : "A JSX-based page builder for creating beautiful websites without writing code"
